pkg/restapi/kms/operation/operations.go

```diff
@@ -172,129 +178,185 @@ func (o *Operation) createKeyHandler(rw http.ResponseWriter, req *http.Request)
 	rw.WriteHeader(http.StatusCreated)
 }
 
-func createKey(provider *kmsProvider, keystoreID, keyType string) (string, error) {
-	// TODO: Pass kms.Service as a dependency (https://github.com/trustbloc/hub-kms/issues/29)
-	srv := kmsservice.NewService(provider)
+func (o *Operation) signHandler(rw http.ResponseWriter, req *http.Request) {
+	var request signReq
+	if ok := parseRequest(&request, rw, req); !ok {
+		return
+	}
+
+	keystoreID := mux.Vars(req)[keystoreIDQueryParam]
+	keyID := mux.Vars(req)[keyIDQueryParam]
+
+	provider := prepareKMSProvider(rw, o.provider, keystoreID, request.Passphrase)
+	if provider == nil {
+		return
+	}
 
-	keyID, err := srv.CreateKey(keystoreID, kms.KeyType(keyType))
+	srv := kmsservice.NewService(provider)
+	signature, err := srv.Sign(keystoreID, keyID, []byte(request.Message))
 	if err != nil {
-		return "", err
+		writeErrorResponse(rw, http.StatusInternalServerError, fmt.Sprintf(signMessageFailure, err))
+		return
 	}
 
-	return keyID, nil
+	writeResponse(rw, signResp{
+		Signature: base64.URLEncoding.EncodeToString(signature),
+	})
 }
 
-func (o *Operation) signHandler(rw http.ResponseWriter, req *http.Request) {
-	var request signReq
-	if err := json.NewDecoder(req.Body).Decode(&request); err != nil {
-		writeErrorResponse(rw, http.StatusBadRequest, fmt.Sprintf(receivedBadRequest, err))
+func (o *Operation) verifyHandler(rw http.ResponseWriter, req *http.Request) {
+	var request verifyReq
+	if ok := parseRequest(&request, rw, req); !ok {
 		return
 	}
 
-	keystoreID := mux.Vars(req)["keystoreID"]
-	keyID := mux.Vars(req)["keyID"]
+	keystoreID := mux.Vars(req)[keystoreIDQueryParam]
+	keyID := mux.Vars(req)[keyIDQueryParam]
 
-	provider, err := prepareKMSProvider(o.provider, KMSCreatorContext{
-		KeystoreID: keystoreID,
-		Passphrase: request.Passphrase,
-	})
+	provider := prepareKMSProvider(rw, o.provider, keystoreID, request.Passphrase)
+	if provider == nil {
+		return
+	}
+
+	signature, err := base64.URLEncoding.DecodeString(request.Signature)
 	if err != nil {
-		writeErrorResponse(rw, http.StatusInternalServerError, fmt.Sprintf(createKMSProviderFailure, err))
+		writeErrorResponse(rw, http.StatusBadRequest, fmt.Sprintf(receivedBadRequest, err))
 		return
 	}
 
-	signature, err := sign(provider, keystoreID, keyID, request.Message)
+	srv := kmsservice.NewService(provider)
+	err = srv.Verify(keystoreID, keyID, signature, []byte(request.Message))
 	if err != nil {
-		writeErrorResponse(rw, http.StatusInternalServerError, fmt.Sprintf(signMessageFailure, err))
+		status := http.StatusInternalServerError
+		if errors.Is(err, kmsservice.ErrInvalidSignature) {
+			status = http.StatusOK
+		}
+
+		writeErrorResponse(rw, status, fmt.Sprintf(verifyMessageFailure, err))
```

Comment:
Is this failure message leaking details? That's not good practice (for security reasons)

Reply:
agree, created follow-up ticket https://github.com/trustbloc/hub-kms/issues/38